File: native-integration/android-plugin/java/com/roxfollow/app/UnityRewardedAdsPlugin.java
package com.roxfollow.app;

import android.app.Activity;
import android.util.Log;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import com.unity3d.ads.IUnityAdsInitializationListener;
import com.unity3d.ads.IUnityAdsLoadListener;
import com.unity3d.ads.IUnityAdsShowListener;
import com.unity3d.ads.UnityAds;
import com.unity3d.ads.UnityAdsShowOptions;

@CapacitorPlugin(name = "UnityRewardedAds")
public class UnityRewardedAdsPlugin extends Plugin {

    private static final String TAG = "UnityRewardedAds";

    private static final String GAME_ID = "800368206";
    private static final String PLACEMENT_ID = "Rewarded_Android";

    // TEST MODE ON
    private static final boolean TEST_MODE = false;

    private boolean initialized = false;
    private boolean loaded = false;
    private PluginCall pendingCall = null;

    @Override
    public void load() {
        super.load();
        initialize();
    }

    private void initialize() {
        Activity activity = getActivity();

        UnityAds.initialize(
            activity.getApplicationContext(),
            GAME_ID,
            TEST_MODE,
            new IUnityAdsInitializationListener() {

                @Override
                public void onInitializationComplete() {
                    initialized = true;
                    Log.d(TAG, "Unity Ads initialized - TEST MODE ON");
                    loadRewarded();
                }

                @Override
                public void onInitializationFailed(
                    UnityAds.UnityAdsInitializationError error,
                    String message
                ) {
                    initialized = false;
                    Log.e(
                        TAG,
                        "Unity Ads initialization failed: "
                        + error + " / " + message
                    );

                    resolve("INITIALIZATION_FAILED:" + message);
                }
            }
        );
    }

    private void loadRewarded() {
        Log.d(TAG, "Loading Rewarded: " + PLACEMENT_ID);

        UnityAds.load(
            PLACEMENT_ID,
            new IUnityAdsLoadListener() {

                @Override
                public void onUnityAdsAdLoaded(String placementId) {
                    loaded = true;

                    Log.d(
                        TAG,
                        "Rewarded ad loaded: " + placementId
                    );

                    // If user already pressed Watch Ad,
                    // show automatically after loading.
                    if (pendingCall != null) {
                        showLoadedRewarded();
                    }
                }

                @Override
                public void onUnityAdsFailedToLoad(
                    String placementId,
                    UnityAds.UnityAdsLoadError error,
                    String message
                ) {
                    loaded = false;

                    Log.e(
                        TAG,
                        "Rewarded load failed: "
                        + error + " / " + message
                    );

                    resolve("LOAD_FAILED:" + error + ":" + message);
                }
            }
        );
    }

    @PluginMethod
    public void isAdReady(PluginCall call) {
        JSObject result = new JSObject();
        result.put("ready", loaded);
        call.resolve(result);
    }

    @PluginMethod
    public void showRewardedAd(PluginCall call) {

        Log.d(
            TAG,
            "showRewardedAd called. initialized="
            + initialized
            + ", loaded="
            + loaded
        );

        if (!initialized) {
            pendingCall = call;

            Log.d(
                TAG,
                "Unity not initialized. Waiting for initialization."
            );

            return;
        }

        if (!loaded) {
            pendingCall = call;

            Log.d(
                TAG,
                "Rewarded not ready. Waiting for ad load."
            );

            loadRewarded();
            return;
        }

        pendingCall = call;
        showLoadedRewarded();
    }

    private void showLoadedRewarded() {

        if (pendingCall == null) {
            return;
        }

        if (!loaded) {
            return;
        }

        loaded = false;

        Log.d(
            TAG,
            "Showing Unity Rewarded: " + PLACEMENT_ID
        );

        UnityAds.show(
            getActivity(),
            PLACEMENT_ID,
            new UnityAdsShowOptions(),

            new IUnityAdsShowListener() {

                @Override
                public void onUnityAdsShowFailure(
                    String placementId,
                    UnityAds.UnityAdsShowError error,
                    String message
                ) {
                    Log.e(
                        TAG,
                        "Rewarded show failed: "
                        + error + " / " + message
                    );

                    resolve(
                        "SHOW_FAILED:"
                        + error
                        + ":"
                        + message
                    );

                    loadRewarded();
                }

                @Override
                public void onUnityAdsShowStart(
                    String placementId
                ) {
                    Log.d(TAG, "Rewarded show started");
                }

                @Override
                public void onUnityAdsShowClick(
                    String placementId
                ) {
                    Log.d(TAG, "Rewarded clicked");
                }

                @Override
                public void onUnityAdsShowComplete(
                    String placementId,
                    UnityAds.UnityAdsShowCompletionState state
                ) {

                    Log.d(
                        TAG,
                        "Rewarded completed. State="
                        + state
                    );

                    if (
                        state
                        == UnityAds.UnityAdsShowCompletionState.COMPLETED
                    ) {
                        resolve("REWARDED");
                    } else {
                        resolve("SKIPPED");
                    }

                    loadRewarded();
                }
            }
        );
    }

    private void resolve(String status) {

        if (pendingCall == null) {
            return;
        }

        JSObject result = new JSObject();
        result.put("status", status);

        pendingCall.resolve(result);
        pendingCall = null;
    }
}

File: native-integration/android-plugin/java/com/roxfollow/app/UnityRewardedAdsPlugin.kt
package com.roxfollow.app

import android.app.Activity
import android.util.Log
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.unity3d.ads.IUnityAdsInitializationListener
import com.unity3d.ads.IUnityAdsLoadListener
import com.unity3d.ads.IUnityAdsShowListener
import com.unity3d.ads.UnityAds
import com.unity3d.ads.UnityAdsShowOptions

/**
 * Bridges the web app's Store screen "Watch Ad" button to a real Unity
 * REWARDED video ad. Interstitial ads are intentionally NOT used here.
 *
 * JS side calls: Capacitor.Plugins.UnityRewardedAds.showRewardedAd()
 * -> resolves { status: "REWARDED" | "SKIPPED" | "FAILED" | "NOT_READY" | "NOT_INITIALIZED" }
 * Coins must only be credited by the caller when status === "REWARDED".
 */
@CapacitorPlugin(name = "UnityRewardedAds")
class UnityRewardedAdsPlugin : Plugin() {

    companion object {
        private const val TAG = "UnityRewardedAds"

        // From Unity Dashboard -> Rox Follow app -> Rewarded placement (Android)
        private const val GAME_ID = "800368206"
        private const val REWARDED_PLACEMENT_ID = "Rewarded_Android"

        // Set to true ONLY while testing with Unity's test ads, then set back to false.
        private const val TEST_MODE = false
    }

    private var isSdkInitialized = false
    private var isAdLoaded = false
    private var pendingCall: PluginCall? = null

    override fun load() {
        super.load()
        initializeUnityAds()
    }

    private fun initializeUnityAds() {
        val act: Activity = activity
        UnityAds.initialize(act.applicationContext, GAME_ID, TEST_MODE, object : IUnityAdsInitializationListener {
            override fun onInitializationComplete() {
                isSdkInitialized = true
                Log.d(TAG, "Unity Ads initialized")
                preloadRewardedAd()
            }

            override fun onInitializationFailed(
                error: UnityAds.UnityAdsInitializationError?,
                message: String?
            ) {
                isSdkInitialized = false
                Log.e(TAG, "Unity Ads init failed: $message")
            }
        })
    }

    private fun preloadRewardedAd() {
        UnityAds.load(REWARDED_PLACEMENT_ID, object : IUnityAdsLoadListener {
            override fun onUnityAdsAdLoaded(placementId: String?) {
                isAdLoaded = true
            }

            override fun onUnityAdsFailedToLoad(
                placementId: String?,
                error: UnityAds.UnityAdsLoadError?,
                message: String?
            ) {
                isAdLoaded = false
                Log.e(TAG, "Rewarded ad failed to load: $message")
            }
        })
    }

    @PluginMethod
    fun isAdReady(call: PluginCall) {
        val ret = JSObject()
        ret.put("ready", isAdLoaded)
        call.resolve(ret)
    }

    @PluginMethod
    fun showRewardedAd(call: PluginCall) {
        if (!isSdkInitialized) {
            val ret = JSObject()
            ret.put("status", "NOT_INITIALIZED")
            call.resolve(ret)
            return
        }

        if (!isAdLoaded) {
            val ret = JSObject()
            ret.put("status", "NOT_READY")
            call.resolve(ret)
            preloadRewardedAd()
            return
        }

        pendingCall = call
        val act: Activity = activity

        UnityAds.show(act, REWARDED_PLACEMENT_ID, UnityAdsShowOptions(), object : IUnityAdsShowListener {
            override fun onUnityAdsShowFailure(
                placementId: String?,
                error: UnityAds.UnityAdsShowError?,
                message: String?
            ) {
                Log.e(TAG, "Rewarded ad show failed: $message")
                isAdLoaded = false
                resolveShow("FAILED")
                preloadRewardedAd()
            }

            override fun onUnityAdsShowStart(placementId: String?) {}

            override fun onUnityAdsShowClick(placementId: String?) {}

            override fun onUnityAdsShowComplete(
                placementId: String?,
                state: UnityAds.UnityAdsShowCompletionState?
            ) {
                isAdLoaded = false
                if (state == UnityAds.UnityAdsShowCompletionState.COMPLETED) {
                    resolveShow("REWARDED")
                } else {
                    resolveShow("SKIPPED")
                }
                preloadRewardedAd()
            }
        })
    }

    private fun resolveShow(status: String) {
        val ret = JSObject()
        ret.put("status", status)
        pendingCall?.resolve(ret)
        pendingCall = null
    }
}
